res/controllers/Traktor-Kontrol-S2-MK3-hid-scripts.js

```diff
@@ -0,0 +1,974 @@
+    message.setCallback(group, name, callback);
+};
+
+TraktorS2MK3.registerInputButton = function (message, group, name, offset, bitmask, callback) {
+    message.addControl(group, name, offset, "B", bitmask);
+    message.setCallback(group, name, callback);
+};
+
+TraktorS2MK3.playHandler = function (field) {
+    if (field.value === 0) {
+        return;
+    }
+
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "cue_set", field.value);
+    } else {
+        var playing = engine.getValue(field.group, "play");
+        engine.setValue(field.group, "play", !playing);
+    }
+};
+
+TraktorS2MK3.cueHandler = function (field) {
+    if (field.value === 0) {
+        return;
+    }
+
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "cue_gotoandstop", field.value);
+    } else {
+        engine.setValue(field.group, "cue_default", field.value);
+    }
+};
+
+TraktorS2MK3.shiftHandler = function (field) {
+    TraktorS2MK3.shiftPressed[field.group] = field.value;
+    engine.setValue("[Controls]", "touch_shift", field.value);
+    TraktorS2MK3.outputHandler(field.value, field.group, "shift");
+};
+
+TraktorS2MK3.keylockHandler = function (field) {
+    if (field.value === 0) {
+        return;
+    }
+
+    var keylock = engine.getValue(field.group, "keylock");
+    engine.setValue(field.group, "keylock", !keylock);
+};
+
+TraktorS2MK3.syncHandler = function (field) {
+    if (TraktorS2MK3.shiftPressed[field.group]) {
+        engine.setValue(field.group, "beatsync_phase", field.value);
+        // Light LED while pressed
+        TraktorS2MK3.outputHandler(field.value, field.group, "sync_enabled");
+    } else {
+        if (field.value) {
+            if (!TraktorS2MK3.syncPressedTimer[field.group]) {
+                // Start timer to measure how long button is pressed
+                TraktorS2MK3.syncPressedTimer[field.group] = engine.beginTimer(1000, "TraktorS2MK3.syncTimer(\"" + field.group + "\")");
+            }
+
+            var sync = engine.getValue(field.group, "sync_enabled");
+            engine.setValue(field.group, "sync_enabled", !sync);
+        } else {
+            // Button is released, check if timer is still running
+            if (TraktorS2MK3.syncPressedTimer[field.group] === 0) {
+                // long press -> sync lock
+                engine.stopTimer(TraktorS2MK3.syncPressedTimer[field.group]);
+                TraktorS2MK3.syncPressedTimer[field.group] = 0;
+            } else {
+                // short press -> disable sync
+                engine.setValue(field.group, "sync_enabled", 0);
+            }
+        }
+    }
+};
+
+TraktorS2MK3.syncTimer = function (group) {
+    // Reset sync button timer if active
+    if (TraktorS2MK3.syncPressedTimer[group] !== 0) {
+        engine.stopTimer(TraktorS2MK3.syncPressedTimer[group]);
+        TraktorS2MK3.syncPressedTimer[group] = 0;
```

Comment:
You're resetting the timer, but you're not changing any control object. According to your wiki entry, the mapping is supposed to toggle `sync_master`.

Reply:
As written in https://www.mixxx.org/wiki/doku.php/mixxxcontrols#channeln under `sync_enabled` Mixxx distinguishes between pressing the SYNC button for a long period (< 1 sec) and a short press. I just imitate this behavior in my script, since Mixxx is not checking the duration when setting the control from a controller script. This is also mentioned in the Wiki page, but i made it more clear there. The timer is started when the button is pressed. When it is released the `syncHandler()` function is checking, if the timer is still running. In this case the button was pressed shorter than one second the SYNC lock is not activated by reseting the control. But you pointed to the right spot, there was some unnecessary code in `syncHandler()`, because stopping the timer when it is already expired is useless. I fixed this in https://github.com/mi01/mixxx/commit/0ac84084748407cab4e81aefad0e6544e351c3f8